mypy/checker.py

```diff
@@ -2482,9 +2482,37 @@ def check_assignment_to_multiple_lvalues(self, lvalues: List[Lvalue], rvalue: Ex
             # using the type of rhs, because this allowed more fine grained
             # control in cases like: a, b = [int, str] where rhs would get
             # type List[object]
-
-            rvalues = rvalue.items
-
+            rvalues = []  # type: List[Expression]
+            lhs_len = len(lvalues)
+            idx_of_iterable = []
+            item_type_of_iterable = []  # type: List['mypy.types.Type']
+            idx = 0
+            for rval in rvalue.items:
+                if isinstance(rval, StarExpr):
+                    typs = get_proper_type(self.expr_checker.visit_star_expr(rval).type)
+                    if isinstance(typs, TupleType):
+                        rvalues.extend([TempNode(typ) for typ in typs.items])
+                        lhs_len -= len(typs.items)
+                        idx += len(typs.items)
+                    elif self.type_is_iterable(typs) and isinstance(typs, Instance):
+                        item_type_of_iterable.append(self.iterable_item_type(typs))
+                        idx_of_iterable.append(idx)
+                    else:
+                        self.fail("StarExpr should not be a '{}'".format(typs), context)
+                else:
+                    rvalues.append(rval)
+                    lhs_len -= 1
+                    idx += 1
+            num_every_iterable = 0
+            num_last_iterable = 0
+            if len(idx_of_iterable):
+                num_every_iterable = int(lhs_len / len(idx_of_iterable))
+                num_last_iterable = lhs_len - (len(idx_of_iterable) - 1) * int(num_every_iterable)
```

Comment:
This seems kind of arbitrary. We assume that the items from multiple `*args` rvalues are split between the values in a specific way, but we can't really predict this. Instead, I'd suggest this:
* If all the item types of iterable `*args` rvalues are the same and next to each other, treat them as a single iterable.
* If the item types vary or they aren't next to each other generate an error and fall back to some conservative approximation, such as assuming all their types are `Any`.

I think that the current behavior can be unsafe. Consider this:

```py
a = [1]
b = ['x', 'y', 'z']
x1, x2, x3, x4 = *a, *b
x2 + 5  # Runtime error
```